File: sample-chat-obj-c/sample-chat/Components/ConnectionModule/ConnectionModule.m
//
//  Connection.m
//  sample-chat
//
//  Created by Injoit on 06.10.2020.
//  Copyright © 2020 QuickBlox Team. All rights reserved.
//

#import "ConnectionModule.h"
#import "Profile.h"
#import <arpa/inet.h>

static NSInteger const ALREADY_CONNECTED_CODE = -1000;

@interface ConnectionModule ()

@property (nonatomic, strong) id appActiveStateObserver;
@property (nonatomic, strong) id appInactiveStateObserver;

@property (nonatomic, assign) SCNetworkReachabilityRef reachability;
@property (nonatomic, assign, readonly) BOOL isNetworkLost;

@end

@interface ConnectionModule (ChatConnection) <QBChatDelegate>
@end

@implementation ConnectionModule

- (SCNetworkReachabilityRef)reachability {
    if (_reachability) {
        return _reachability;
    }
    struct sockaddr_in address;
    bzero(&address, sizeof(address));
    address.sin_len = sizeof(address);
    address.sin_family = AF_INET;
    
    _reachability =
    SCNetworkReachabilityCreateWithAddress(kCFAllocatorDefault, (const struct sockaddr *)&address);
    return _reachability;
}

- (BOOL)isNetworkLost {
    SCNetworkReachabilityFlags flags;
    if (SCNetworkReachabilityGetFlags(self.reachability, &flags)) {
        
        return (flags & kSCNetworkReachabilityFlagsReachable) == 0;
    }
    return NO;
}

- (id)appActiveStateObserver {
    if (_appActiveStateObserver) {
        return _appActiveStateObserver;
    }
    NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
    __weak __typeof(self)weakSelf = self;
    _appActiveStateObserver = [center addObserverForName:UIApplicationWillEnterForegroundNotification
                                                  object:nil
                                                   queue:NSOperationQueue.mainQueue
                                              usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf establish];
    }];
    return _appActiveStateObserver;
}

- (id)appInactiveStateObserver {
    if (_appInactiveStateObserver) {
        return _appInactiveStateObserver;
    }
    NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
    __weak __typeof(self)weakSelf = self;
    _appInactiveStateObserver = [center addObserverForName:UIApplicationDidEnterBackgroundNotification
                                                  object:nil
                                                   queue:NSOperationQueue.mainQueue
                                              usingBlock:^(NSNotification * _Nonnull note) {
        [weakSelf disconnect];
    }];

    return _appInactiveStateObserver;
}

- (BOOL)tokenHasExpired {
    return QBSession.currentSession.tokenHasExpired;
}

- (BOOL)established {
    BOOL connected = NO;
    if (self.tokenHasExpired == NO && QBChat.instance.isConnected) {
        connected = YES;
    }
    return connected;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        QBSettings.autoReconnectEnabled = YES;
        QBSettings.networkIndicatorManagerEnabled = YES;
        [QBChat.instance addDelegate:self];
    }
    return self;
}

- (void)dealloc {
    [QBChat.instance removeDelegate:self];
}

//MARK: - Actions
- (void)activateAutomaticMode {
    [self appActiveStateObserver];
    
    [self appInactiveStateObserver];
    
    [self establish];
}

- (void)deactivateAutomaticMode {
    NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
    [center removeObserver:self.appActiveStateObserver];
    [center removeObserver:self.appInactiveStateObserver];
}

- (void)establish {
    if (self.tokenHasExpired ) {
        [self disconnect];
        if ([self.delegate respondsToSelector:@selector(connectionModuleTokenHasExpired:)]) {
            [self.delegate connectionModuleTokenHasExpired:self];
        }
        return;
    }
    if (QBChat.instance.isConnected) {
        if ([self.delegate respondsToSelector:@selector(connectionModuleDidConnect:)]) {
            [self.delegate connectionModuleDidConnect:self];
        }
        return;
    }
    
    if ([self.delegate respondsToSelector:@selector(connectionModuleWillConnect:)]) {
        [self.delegate connectionModuleWillConnect:self];
    }
    
    __weak __typeof(self)weakSelf = self;
    [QBChat.instance connectWithUserID:QBSession.currentSession.sessionDetails.userID
                              password:QBSession.currentSession.sessionDetails.token
                            completion:^(NSError * _Nullable error) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (error && error.code != ALREADY_CONNECTED_CODE) {
            if ([strongSelf.delegate respondsToSelector:@selector(connectionModuleDidNotConnect:withError:)]) {
                [strongSelf.delegate connectionModuleDidNotConnect:strongSelf withError:error];
            }
            return;
        }
        
        if ([strongSelf.delegate respondsToSelector:@selector(connectionModuleDidConnect:)]) {
            [strongSelf.delegate connectionModuleDidConnect:strongSelf];
        }
    }];
}

- (void)breakConnectionWithCompletion:(nonnull void (^)(void))completion {
    if (QBChat.instance.isConnected == NO) {
        completion();
        return;
    }
    [QBChat.instance disconnectWithCompletionBlock:^(NSError * _Nullable error) {
        completion();
    }];
}

//MARK: - Internal
- (void)disconnect {
    if (QBChat.instance.isConnected == NO) {
        return;
    }
    [QBChat.instance disconnectWithCompletionBlock:nil];
}

@end

@implementation ConnectionModule (ChatConnection)

- (void)chatDidNotConnectWithError:(NSError *)error {
    if ([self.delegate respondsToSelector:@selector(connectionModuleDidNotConnect:withError:)]) {
        [self.delegate connectionModuleDidNotConnect:self withError:error];
    }
}

- (void)chatDidDisconnectWithError:(NSError *)error {
    if (error == nil) {
        return;
    }
    if ([self.delegate respondsToSelector:@selector(connectionModuleDidNotConnect:withError:)]) {
        [self.delegate connectionModuleDidNotConnect:self withError:error];
    }
}

- (void)chatDidReconnect {
    if ([self.delegate respondsToSelector:@selector(connectionModuleDidReconnect:)]) {
        [self.delegate connectionModuleDidReconnect:self];
    }
}

- (void)chatDidAccidentallyDisconnect {
    if ([self.delegate respondsToSelector:@selector(connectionModuleWillConnect:)]) {
        [self.delegate connectionModuleWillConnect:self];
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/ConnectionModule/ConnectionModule.swift
//
//  ConnectionModule.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 02.04.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import QuickbloxWebRTC
import SystemConfiguration

struct ConnectionConstant {
    static let connectingState = "Still in connecting state, please wait"
    static let reconnectingState = "Reconnecting state, please wait"
    static let connectionEstablished = "Connection established"
    static let noInternetConnection = "No Internet Connection"
}

typealias CompletionAction = (() -> Void)

@objc protocol ConnectionModuleDelegate: AnyObject {
    @objc optional func connectionModuleWillConnect(_ connectionModule: ConnectionModule)
    @objc optional func connectionModuleDidConnect(_ connectionModule: ConnectionModule)
    @objc optional func connectionModuleDidNotConnect(_ connectionModule: ConnectionModule, error: Error)
    @objc optional func connectionModuleWillReconnect(_ connectionModule: ConnectionModule)
    @objc optional func connectionModuleDidReconnect(_ connectionModule: ConnectionModule)
    @objc optional func connectionModuleTokenHasExpired(_ connectionModule: ConnectionModule)
}

/// The module is responsible for automatic connection establishment with QuickBlox.
///
/// The connection depends on the application state and call activity.
/// The establishing connection process can start from the authorization if the "QBSession" token was expired.
class ConnectionModule: NSObject {
    //MARK: - Properties
    weak var delegate: ConnectionModuleDelegate?
    private var appActiveStateObserver: Any?
    /// Determining the connection state.
    ///
    /// Calling this property starts the connection process when it's not established.
    var established: Bool {
        var connected = false
        if tokenHasExpired == false, QBChat.instance.isConnected {
            connected = true
        }
        return connected
    }
    /// Determining the "QBSession" token state.
    var tokenHasExpired: Bool {
        return QBSession.current.tokenHasExpired
    }
    
    private var isReachability: SCNetworkReachability? {
        return reachability()
    }
    private var isNetwork: Bool? {
        var flags: SCNetworkReachabilityFlags = []
        guard let reachability = isReachability,
              SCNetworkReachabilityGetFlags(reachability, &flags) else {
            return nil
        }
        return flags.contains(.reachable)
    }
    
    private var activeCall: Bool = false
    
    
    //MARK: - Life Cycle
    override init() {
        super.init()
        
        QBSettings.autoReconnectEnabled = true
        QBSettings.networkIndicatorManagerEnabled = true
        QBChat.instance.addDelegate(self)
    }
    
    deinit {
        QBChat.instance.removeDelegate(self)
    }
    
    // MARK: - Public Methods
    func reachability() -> SCNetworkReachability? {
        var zeroAddress = sockaddr_in()
        zeroAddress.sin_len = UInt8(MemoryLayout<sockaddr_in>.size)
        zeroAddress.sin_family = sa_family_t(AF_INET)
        
       let isReachability = withUnsafePointer(to: &zeroAddress, {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                SCNetworkReachabilityCreateWithAddress(nil, $0)
            }
       })
        return isReachability
    }
    
    func setupAppActiveStateObserver() {
        if appActiveStateObserver != nil {
            return
        }
        let center = NotificationCenter.default
        appActiveStateObserver = center.addObserver(forName: UIApplication.willEnterForegroundNotification,
                                                    object: nil,
                                                    queue: OperationQueue.main,
                                                    using: { [weak self] (note) in
            guard let self = self else { return }
            self.establish()
        })
    }
    
    /// Activating the automatic connection and disconnection process when the application state change.
    ///
    /// Calling this method starts the connection process when it's not established.
    func activateAutomaticMode() {
        setupAppActiveStateObserver()
        establish()
    }
    
    /// Stop trying connection automatically.
    func deactivateAutomaticMode() {
        NotificationCenter.default.removeObserver(appActiveStateObserver as Any)
    }
    
    /// Prevents breaking connection when the application state change.
    func activateCallMode() {
        activeCall = true
    }
    
    /// Break connection when application state is inactive.
    func deactivateCallMode() {
        activeCall = false
    }
    
    /// Establishes a connection with the Quickblox.
     func establish() {
        guard tokenHasExpired == false,
              let sessionToken = QBSession.current.sessionDetails?.token,
              let userId = QBSession.current.sessionDetails?.userID else {
            disconnect()
            delegate?.connectionModuleTokenHasExpired?(self)
            return
        }
        
        if QBChat.instance.isConnected || QBChat.instance.isConnecting {
            delegate?.connectionModuleDidConnect?(self)
            return
        }
        
        delegate?.connectionModuleWillConnect?(self)
        
        QBChat.instance.connect(withUserID: userId, password: sessionToken) { [weak self] (error) in
            guard let self = self else {
                return
            }
            if let error = error, error._code != ErrorCode.alreadyConnectedCode {
                self.delegate?.connectionModuleDidNotConnect?(self, error: error)
                return
            }
            self.delegate?.connectionModuleDidConnect?(self)
        }
    }
    
    /// Disconnects and unauthorize from the Quickblox.
    func breakConnection(withCompletion completion:@escaping CompletionAction) {
        if QBChat.instance.isConnected == false {
            completion()
            return
        }
        
        QBChat.instance.disconnect { (error) in
            completion()
        }
    }
    
    //MARK: - Internal
    private func disconnect() {
        if QBChat.instance.isConnected == false { return }
        QBChat.instance.disconnect(completionBlock: nil)
    }
}

//MARK: - QBChatDelegate
extension ConnectionModule: QBChatDelegate {
    func chatDidNotConnectWithError(_ error: Error) {
        delegate?.connectionModuleDidNotConnect?(self, error: error)
    }
    
    func chatDidDisconnectWithError(_ error: Error?) {
        if let error = error,
           error._code != ErrorCode.socketClosedRemote,
           error._code != ErrorCode.brokenPipe {
            delegate?.connectionModuleDidNotConnect?(self, error: error)
        }
    }
    
    func chatDidAccidentallyDisconnect() {
        delegate?.connectionModuleWillReconnect?(self)
    }
    
    func chatDidReconnect() {
        delegate?.connectionModuleDidReconnect?(self)
    }
}
